--- a/lib/password.c
+++ b/lib/password.c
@@ -11,8 +11,10 @@
 int get_salt(char *salt, char *algo, int rand)
 {
   struct {
-    char *type, id, len;
-  } al[] = {{"des", 0, 2}, {"md5", 1, 8}, {"sha256", 5, 16}, {"sha512", 6, 16}};
+    char *type, *id;
+    int len;
+  } al[] = {{"des", 0, 2}, {"md5", "1", 8}, {"sha256", "5", 16},
+            {"sha512", "6", 16}, {"bcrypt", "2b", 22}};
   char *s;
   int i, len;
 
@@ -24,7 +26,9 @@
     if (algo[len]) continue;
 
     len = al[i].len;
-    s = salt + (al[i].id ? sprintf(salt, "$%c$", '0'+al[i].id) : 0);
+    if (!al[i].id) s = salt;
+    else if (!strcmp(al[i].id, "2b")) s = salt + sprintf(salt, "$2b$05$");
+    else s = salt + sprintf(salt, "$%s$", al[i].id);
 
     // Read appropriate number of random bytes for salt
     if (rand) xgetrandom(libbuf, ((len*6)+7)/8);
@@ -130,7 +134,7 @@
   long long ll = 0;
 
   // Open old filename ("r" won't let us lock) and get blocking lock
-  if (!(ofp = fopen(filename, "w+")) || 0>fcntl(fileno(ofp), F_SETLK, &lock)\
+  if (!(ofp = fopen(filename, "r+")) || 0>fcntl(fileno(ofp), F_SETLK, &lock)\
       || fstat(fileno(ofp), &st))
   {
     perror_msg_raw(filename);
@@ -153,7 +157,7 @@
     // find matching line
     oo = 0;
     start = end = chomp(line);
-    if (strncmp(line, username, len) || !(line[len] && line[len]!=':'))
+    if (strncmp(line, username, len) || line[len] != ':')
       out = line;
     else {
       found++;
@@ -168,8 +172,8 @@
           while (*end && *end != ':') end++;
           if (!pos || !*end) break;
         }
-        if (pos>=0) out = line;
-        else oo = out = xmprintf("%*s%s%s\n", (int)(start-line),line,entry,end);
+        if (pos>0) out = line;
+        else oo = out = xmprintf("%.*s%s%s\n", (int)(start-line),line,entry,end);
       }
     }
     if (out) {
@@ -176,5 +180,4 @@
       rc = dprintf(nfd, "%s\n", out);
-      free(out);
       if (rc<0) {
         perror_msg_raw(ff);
         goto free_storage;
