--- a/lib/password.c
+++ b/lib/password.c
@@ -130,7 +130,7 @@
   long long ll = 0;

   // Open old filename ("r" won't let us lock) and get blocking lock
-  if (!(ofp = fopen(filename, "w+")) || 0>fcntl(fileno(ofp), F_SETLK, &lock)\
+  if (!(ofp = fopen(filename, "r+")) || 0>fcntl(fileno(ofp), F_SETLK, &lock)\
       || fstat(fileno(ofp), &st))
   {
     perror_msg_raw(filename);
@@ -153,7 +153,7 @@
     // find matching line
     oo = 0;
     start = end = chomp(line);
-    if (strncmp(line, username, len) || !(line[len] && line[len]!=':'))
+    if (strncmp(line, username, len) || line[len] != ':')
       out = line;
     else {
       found++;
@@ -168,8 +168,8 @@
           while (*end && *end != ':') end++;
           if (!pos || !*end) break;
         }
-        if (pos>=0) out = line;
-        else oo = out = xmprintf("%*s%s%s\n", (int)(start-line),line,entry,end);
+        if (pos>0) out = line;
+        else oo = out = xmprintf("%.*s%s%s\n", (int)(start-line),line,entry,end);
       }
     }
     if (out) {
@@ -176,5 +176,4 @@
       rc = dprintf(nfd, "%s\n", out);
-      free(out);
       if (rc<0) {
         perror_msg_raw(ff);
         goto free_storage;
